Blog post: Design principles for AiiDAlab apps (#77)
aiidateam/aiida-website: docs/news/posts/2024-11-15-aiidalab-mvc.md

        self.textbox = ipw.Text(description="Label")
        self.button = ipw.Button(
            description="Submit",
            button_style="success",
            icon="check",
        )
        super().__init__(
            children=[
                self.dropdown,
                ipw.HBox(
                    children=[
                        self.textbox,
                        self.button,
                    ],
                ),
            ],
        )
```

![View example]()

The class represents a UI for the user to interact with the app. It typically provides an interface for the user to select and provide input for the underlying simulation software.

**Tip 💡** consider arranging UI components in a pattern closely representative of the flow of app operations. This is often best determined hand-in-hand with the intended user base.

#### `View`-`Model` synchronization

The observable state variables defined in the `Model` can be leveraged to synchronize the `View` with the current state of the app. This can be done via the `traitlets` linking mechanism. This can be set up in the `Controller` class.

### The `Controller`

> The purpose of the `Controller` class, the final component of the MVC design patter, is to handle events and control the flow of information across the app. This means reacting to changes in models and/or user-driven view events, updating either according to the defined app logic. To do so, a `Controller` is typically instantiated with instances of both the `View` and the `Model`, as follows:

Reviewer: ```suggestion
The purpose of the `Controller` class, the final component of the MVC design pattern, is to handle events and control the flow of information across the app. This means reacting to changes in models and/or user-driven view events, updating according to the defined app logic. To do so, a `Controller` is typically instantiated with instances of both the `View` and the `Model`, as follows:
```

Author: Good catch on "patter", but why take "either" out? The `Controller` updates EITHER the model or the view 🤔

Reviewer: Ah, OK, now I get it. I read it as "...updating either according to the defined app logic <_or something else_>.", didn't realize the _either_ referred to the sentence **before**. Maybe add a comment afterwards to make it more clear.

For 1 and 3, dedicated `_on_event` methods are a good practice for readability, each encapsulating one or more operations (methods) to perform on either the `View` or the `Model`.

## The model network

The models of the various MVC components form the backbone of the app. To connect the models, once can choose one or both of the following patterns.

### The Observer pattern

In the observer pattern, traits of one MVC `Model` are linked to a copy in another. When the source trait changes, the other is synchronized. The `Controller` of the latter `Model` observes the change and update its local MVC component, thus facilitating the flow of data in the app.

### The Mediator pattern

On can also choose to opt for more granular control. In the Mediator pattern, traits of one `MVC` pattern are observed locally by its `Controller`, with the `Controller` subsequently taking action on another `Model`. In such cases, the latter `Model` is often of a dependent MVC component, of which the `Controller` is aware.

---

**Tip 💡** typically, it is not a matter of choosing one or the other, but instead balancing the two patterns to best fit the logic of the app.

## Services and backends

> As apps grow, providing additional features may come with changes in core mechanics. In anticipation of such changes, it is often best to isolate the core mechanics of an app in one or more dedicated services, to be injected into models on demend.

Reviewer: ```suggestion
As apps grow, providing additional features may come with changes in core mechanics. In anticipation of such changes, it is often best to isolate the core mechanics of an app in one or more dedicated services, to be injected into models on demand:
```

Author: Hmm, I wouldn't put a colon at the end there. It's not directly referencing the following snippet.

```python

class AiiDAService:
    def get_process_uuid(self):
        return "..."
```

```python

import traitlets as tl

class Model(tl.HasTraits):
    process_uuid = tl.Unicode(...)

    def __init__(self, aiida_service):

>         self._aiida = aiida_service

Reviewer: ```suggestion
        self._aiida_service = aiida_service
```
?

Author: Here I'm referring to the service as `_aiida`, to give the appearance that I am asking something FROM AiiDA.

    def update(self): ...

    def reset(self):
        self.process_uuid = self._aiida.get_process_uuid()
```

> The above demonstarate how common functionality such as AiiDA interactions could be encapsulated in an `AiiDAService`. When AiiDA core mechanics change, **app maintenance is reduced** to modifications of the service. Common uses of services include database and/or network interactions, logging, utilities, and more.

Reviewer: ```suggestion
The above demonstrates how common functionality such as AiiDA interactions could be encapsulated in an `AiiDAService`. When AiiDA core mechanics change, **app maintenance is reduced** to modifications of the service. Common uses of services include database and/or network interactions, logging, utilities, and more.
```

> #### Exchangable backends

Reviewer: ```suggestion
### Exchangable backends
```

```python

class View(ipw.VBox):

    def __init__(self, **kwargs):
        super().__init__(
            children=[
                LoadingWidget("Loading the view"),  # or any custom spinner widget
                **kwargs,
            ]
        )

        self.rendered = False

    def render(self):
        if self.rendered:
            return

        self.dropdown = ipw.Dropdown(description="Options")
        self.textbox = ipw.Text(description="Label")
        self.button = ipw.Button(
            description="Submit",
            button_style="success",
            icon="check",
        )
        children=[
            self.dropdown,
            ipw.HBox(
                children=[
                    self.textbox,
                    self.button,
                ],
            ),
        ]

        self.rendered = True
```

An associated `Controller` can handle the rendering of the `View` (calling `render()`) at the precise moment it is visited by the user, thus deferring the loading impact to a later stage of the app.

> **Tip 💡** to learn more, you can check out [this PR](https://github.com/aiidalab/aiidalab-qe/pull/802) implementing lazy loading in the Quantum ESPRESSO app.

Reviewer: "Also, feel free to leave a review, as no-one in the `aiidateam` wants to go through 10k lines of changed code" :D
Wouldn't call this "Tip", though. Maybe instead add a footnote?

Author: Right, footnote would be better. Also, while we're on the subject, feel free to review the PR :)

### Concurrency and parallelization

It will often be the case that some operations of your app will take longer than others. To reduce the impact on the experience of the user, one should consider asynchronous operations (concurrency) and/or threading and other parallelization techniques. Here we reference you to a great article about [burgers](https://tiangolo.medium.com/concurrent-burgers-understand-async-await-eeec05ae7cfe).

**Warning** ⚠️ if you do choose to leverage threading, try to avoid sharing AiiDA node instances between threads. In other words, one thread should avoid making use of a node from another thread. Often, this will lead to DB session conflicts. You can read more about it [here](https://aiida.discourse.group/t/using-thread-to-watch-workchains-status-in-aiidalab/135).

## Final thoughts

> Plan your design! The impact early design choices will have on your app down the road are hard to measure in advance. But solid design plan leveraging standard patterns can at least provide some assurance that future changes will require minimal (isolated) maintenance. We encourage you to read more about design patterns in general. A great resource is [Refactoring.Guru](https://refactoring.guru/design-patterns), where you can find examples in many programming languages.

Reviewer: ```suggestion
Plan your design! The impact early design choices will have on your app down the road are hard to measure in advance. But a solid design plan leveraging standard patterns can at least provide some assurance that future changes will require minimal (isolated) maintenance. We encourage you to read more about design patterns in general. A great resource is [Refactoring.Guru](https://refactoring.guru/design-patterns), where you can find examples in many programming languages.
```

Also, be sure to visit the [AiiDA plugin registry](https://aiidateam.github.io/aiida-registry/) to check out the available plugins you could build apps for today! And of course, you can build your own AiiDA plugin and companion app for use in your lab and/or research. We hope you find our tools useful 🙂 And please feel free to reach out to us anytime on [Discourse](https://aiida.discourse.group/) if you have any questions.

---

> And as always, happy computing 🎉

Reviewer: ```suggestion
As always, happy computing 🎉
```
"And" appears quite a few times in the last sentences.
